src/range.rs

```diff
@@ -39,47 +34,84 @@ impl fmt::Debug for TextRange {
     }
 }
 
-/// Creates a new `TextRange` with given `start` and `end.
+/// Creates a new `TextRange` with the given `start` and `end` (`start..end`).
 ///
 /// # Panics
 ///
 /// Panics if `end < start`.
 #[allow(non_snake_case)]
+#[inline]
 pub fn TextRange(start: TextSize, end: TextSize) -> TextRange {
     assert!(start <= end);
     TextRange { start, end }
 }
 
-/// Identity methods.
 impl TextRange {
-    /// Creates a zero-length range at the specified offset.
-    pub const fn empty(self, offset: TextSize) -> TextRange {
+    /// Create a zero-length range at the specified offset (`offset..offset`).
+    #[inline]
+    pub const fn empty(offset: TextSize) -> TextRange {
         TextRange {
             start: offset,
             end: offset,
         }
     }
 
+    /// Create a range up to the given end (`..end`).
+    #[inline]
+    pub const fn before(end: TextSize) -> TextRange {
+        TextRange {
+            start: TextSize::zero(),
+            end,
+        }
+    }
+
+    /// Create a range after the given start (`start..`).
+    ///
+    /// This returns a std [`RangeFrom`] rather than `TextRange` because
+    /// `TextRange` does not support right-unbounded ranges. As such, this
+    /// should only be used for direct indexing, and bounded ranges should be
+    /// used for persistent ranges (`TextRange(start, TextSize::of(text))`).
+    #[inline]
+    pub const fn after(start: TextSize) -> RangeFrom<usize> {
+        start.raw as usize..
+    }
+
+    /// Offset this range by some amount.
+    ///
+    /// This is typically used to convert a range from one coordinate space to
+    /// another, such as from within a substring to within an entire document.
+    #[inline]
+    pub fn offset(self, offset: TextSize) -> TextRange {
```

Comment:
To my mind, `range +- offset` is much more self-explanatory than `range.offset(offset)`.

Reply:
At the very least, a dedicated function provides a clear place to document potential footguns, such as how it can't be used to slide the `TextRange` within a substring, only to change what base it's measured from.

Additionally, I'd expect `Add::add` implementations to respect the customary debug-checked, release-wrapped behavior, which, in order to preserve the `start <= end` invariant, this operation cannot.

Reply:
I still see no reason why the calling code would do `range.offset(offset)` over `range + offset`.

Reply:
Well for one we haven't re-added `range + offset` yet 😄